util/chunk/chunk_util.go

```diff
@@ -0,0 +1,105 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+// CopySelectedJoinRows copies the selected joined rows from the source Chunk
+// to the destination Chunk.
+//
+// NOTE: All the outer rows in the source Chunk should be the same.
+func CopySelectedJoinRows(src *Chunk, innerColOffset, outerColOffset int, selected []bool, dst *Chunk) bool {
+	if src.NumRows() == 0 {
+		return false
+	}
+
+	numSelected := copySelectedInnerRows(innerColOffset, outerColOffset, src, selected, dst)
+	copyOuterRows(innerColOffset, outerColOffset, src, numSelected, dst)
+	dst.numVirtualRows += numSelected
+	return numSelected > 0
+}
+
+// copySelectedInnerRows copies the selected inner rows from the source Chunk
+// to the destination Chunk.
+func copySelectedInnerRows(innerColOffset, outerColOffset int, src *Chunk, selected []bool, dst *Chunk) int {
+	oldLen := dst.columns[innerColOffset].length
+	var srcCols []*column
+	if innerColOffset == 0 {
+		srcCols = src.columns[:outerColOffset]
+	} else {
+		srcCols = src.columns[innerColOffset:]
+	}
+	for j, srcCol := range srcCols {
+		dstCol := dst.columns[innerColOffset+j]
+		if srcCol.isFixed() {
+			for i := 0; i < len(selected); i++ {
+				if !selected[i] {
+					continue
+				}
+				dstCol.appendNullBitmap(!srcCol.isNull(i))
+				dstCol.length++
+
+				elemLen := len(srcCol.elemBuf)
+				offset := i * elemLen
+				dstCol.data = append(dstCol.data, srcCol.data[offset:offset+elemLen]...)
+			}
+		} else {
+			for i := 0; i < len(selected); i++ {
+				if !selected[i] {
+					continue
+				}
+				dstCol.appendNullBitmap(!srcCol.isNull(i))
+				dstCol.length++
+
+				start, end := srcCol.offsets[i], srcCol.offsets[i+1]
+				dstCol.data = append(dstCol.data, srcCol.data[start:end]...)
+				dstCol.offsets = append(dstCol.offsets, int32(len(dstCol.data)))
+			}
+		}
+	}
+	return dst.columns[innerColOffset].length - oldLen
+}
+
+// copyOuterRows copies the continuous 'numRows' outer rows in the source Chunk
+// to the destination Chunk.
+func copyOuterRows(innerColOffset, outerColOffset int, src *Chunk, numRows int, dst *Chunk) {
+	if numRows <= 0 {
+		return
+	}
+	row := src.GetRow(0)
+	var srcCols []*column
+	if innerColOffset == 0 {
+		srcCols = src.columns[outerColOffset:]
+	} else {
+		srcCols = src.columns[:innerColOffset]
+	}
+	for i, srcCol := range srcCols {
+		dstCol := dst.columns[outerColOffset+i]
+		dstCol.appendMultiSameNullBitmap(!srcCol.isNull(row.idx), numRows)
+		dstCol.length += numRows
+		if srcCol.isFixed() {
+			elemLen := len(srcCol.elemBuf)
+			start := row.idx * elemLen
+			end := start + numRows*elemLen
+			dstCol.data = append(dstCol.data, srcCol.data[start:end]...)
+		} else {
+			start, end := srcCol.offsets[row.idx], srcCol.offsets[row.idx+numRows]
+			dstCol.data = append(dstCol.data, srcCol.data[start:end]...)
+			offsets := dstCol.offsets
+			elemLen := srcCol.offsets[row.idx+1] - srcCol.offsets[row.idx]
```

Comment:
Will it be possible that `row.Idx+1` out of range?

Reply:
In the normal case, it will not out of range.

Reply:
Will it be possible that `src.NumRows` equals 1, if so, it seems this will be out of range? @crazycs520

Reply:
if `src.NumRows` equals 1,  the lengh of `offsets` should be 2.

Reply:
ok